### ISSUE TITLE
[Feature request] New rule: no-unnecessary-arbitrary-value

### ISSUE THREAD

**Issue by threehams** [francoismassart/eslint-plugin-tailwindcss#175]
**Is your feature request related to a problem? Please describe.**
I'm working on a team of about 30 people on a medium-sized codebase. It's common to see an arbitrary value like this:
```
text-[14px]
```
This is exactly the same as:
```
text-md
```
However, it creates duplicate CSS. Over time, this adds up to a significant amount of extra CSS for no benefit.

**Describe the solution you'd like**
Detect when an arbitrary value is identical to a configured value, and offer to replace it.

```
bg-[#f8f8f8]
// fixes to
bg-light
```

```
text-[var(--gray)]
// fixes to
text-dark
```

This is all dependent on the project's configuration.

Out of scope:
- Similar (but not exact) colors are not flagged. Only exact matches are done.
- No unit conversion is done to/from rem/px, for example.

**Describe alternatives you've considered**
1. The current rule is `no-arbitrary-value`, but that shuts them off completely.
2. More granularity on `no-arbitrary-value`. There are specific things that tend to get duplicated, like `text-`, `bg-`, and `border-`. Disabling arbitrary values on these could help prevent these from spreading.
3. Instead of going completely for the "all values", make a very targeted rule for the most common offenders.

**Risks:
Arbitrary values come later in the stylesheet, so a fixer will not be 100% safe to run. However, these should already be considered bugs, hopefully caught by `no-contradicting-classname`.

**Additional context**
I'm working on this right now for a private project, so I'll get it done regardless in the next couple weeks, but I thought it may be useful to have it more broadly available.

**Comment by francoismassart** [comment]
Good idea 👍

**Comment by francoismassart** [comment]
@threehams, the last example you listed might be incorrect:

```
text-[var(--gray)]
// fixes to
text-dark
```

The variable could hold a different/updated value based on the css variable usage.
For example a product page using a themed text color, by redeclaring the --var

**Comment by threehams** [comment]
I left out the Tailwind config, which would look like this:
```
textColor: {
  dark: "var(--gray-800)"
}
```
CSS vars are only swapped for config when the config resolves to the CSS var text - it won't try to go find the value of a variable.

**Comment by martinsjek** [comment]
Any updates on this?

**Comment by threehams** [comment]
It's been backlogged at work because:

- There are simpler and quicker ways to reduce CSS size (which is the main benefit for us here).
- I had a lot of difficulty understanding this repo, in particular how to look up values from config. There are many similar rules, but nothing that overlaps enough to make this an easy task.

This doesn't mean I won't work on it at all, just that other things are piled up in front of it for now.

Anyone else interested in working on this wouldn't conflict with anything I've done so far.

**Comment by KMavr** [comment]
Hey, I ran into the same need and ended up building a standalone plugin rather than waiting: [eslint-plugin-tailwind-design-tokens](https://github.com/KMavr/eslint-plugin-tailwind-design-tokens)

The no-unnecessary-arbitrary-value rule covers the general case, but if anyone's looking specifically for color-token-aware detection (i.e. bg-[#f8f8f8] → bg-light based on your design system tokens), the no-hardcoded-colors rule does exactly what's described here, it normalizes arbirtary color values, looks them up against your token map, and either suggests the token name (with autofix when there is an exact match) or flags it as an unexpected hardcoded color when there isn't one. It reads tokens from a Tailwind v4 @theme CSS block, a v3 tailwind.config.js, or an inline map.

**Comment by francoismassart** [comment]
The rule has been migrated to the latest alpha versions.

https://github.com/francoismassart/eslint-plugin-tailwindcss/blob/alpha/v4/docs/rules/no-unnecessary-arbitrary-value.md

At the moment, I'm improving the perfs and then the last step is to move support for ESLint v10

**Comment by KMavr** [comment]
Nice, glad it's landed. Looks like it covers the general redundancy case well (arbitrary value -> preset equivalent).
For anyone landing here specifically on the color angle the plugin I linked takes a slightly different tack: it also flags hardcoded colors that don't have a token eqiuvalent (i.e. enforcing a design system, not just deduplicating against the config), and reads tokens from a v4 @theme block or an inline map rather than only tailwind.config. 
Different scope, might suit anyone whose tokens live in CSS rather than the Tailwind config.
Either way, good to see this in alpha.
